### PR TITLE
michaels work

### PR DESCRIPTION
**Pull request by mayankadlakha**

### CODE REVIEW

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
+using System;
+using Xunit;
+
+namespace StringCalculator.Tests
+{
+    public class StringCalculatorShould
+    {
+        [Fact]
+        public void ReturnZeroFromEmptyString()
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:9]: Could use underscore to make method names more readable `ReturnZero_FromEmptyString` or `EmptyString_ReturnsZero`

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        [Fact]
+        public void ReturnSumFromTwoNumber()
+        {
+            var calcualtor = new StringCalculator();
+            var expected1 = 3;
+            var expected2 = 8;
+
+            var actual1 = calcualtor.Add("1,2");
+            var actual2 = calcualtor.Add("3,5");
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:41]: Testing "1,2" and "3,5" doesn't give us more confidence, so just having one of these test cases should cover the requirement.

Comment (mayankadlakha) [reply StringCalculator.Tests/StringCalculatorTests.cs:41]: Additionally, I would include a test case for "negative, positive". In this case, I would separate the test methods by the type of input. So, `ReturnSumFromTwoPositiveNumbers`, `ReturnExceptionIfNegativeNumber`.

File: StringCalculator/StringCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                var delimiterList = CreateDelimiterList();
+                string[] delimiterArray = delimiterList.ToArray();
+                total = SumStringNumbers(stringInput, total, delimiterArray);
+            }
+
+           return total;
+        }
+
+        public int SumStringNumbers(string text, int total, string[] delimiterArray)
```
Comment (mayankadlakha) [inline StringCalculator/StringCalculator.cs:36]: All these internal methods can be made private. Ideally, you want all your public methods tested.

File: StringCalculator/StringCalculator.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+namespace StringCalculator
+{
+    public class StringCalculator
+    {
+        public StringCalculator()
+        {
+
+        }
+
+        public int Add(string stringInput)
```
Comment (mayankadlakha) [inline StringCalculator/StringCalculator.cs:13]: Add in this case can be a static method and you don't need to initialise the StringCalculator

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        [Fact]
+        public void SupportDifferentDelimiters()
+        {
+            var calculator = new StringCalculator();
+            var expectedDelimiter = ";";
+            var expected = 3;
+
+            var actualDelimiter = calculator.CreateDelimiterList("//;\n1;2")[2];
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:82]: It's good that you are trying to test the `CreateDelimiterList`, however, since it is a private method, you don't need a separate assertion for it. It is implicitly tested in `Add` method's test.

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            var calculator = new StringCalculator();
+            Assert.Throws<ArgumentException>(() => calculator.Add("-1,2,-3"));
+
+            try
+            {
+                calculator.Add("-1,2,-3");
+            }
+            catch(ArgumentException e)
+            {
+                var expected = "Throws exception with Negatives not allowed: -1, -3";
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:101]: The exception message does not need to say "Throws exception", that is something the compiler would add in front of the message by default.

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+
+        [Fact]
+        public void IgnoreNumbersGreaterThanOrEqualTo1000()
+        {
+            var calculator = new StringCalculator();
+            var expected = 2;
+
+            var actual = calculator.Add("1000,1001,2");
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:114]: Nice test case. In these cases where you have a boundary, you also want to test just below the boundary, which would be 999. The first section [here](https://www.guru99.com/equivalence-partitioning-boundary-value-analysis.html) has a nice diagram.

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+        [Fact]
+        public void ReturnSumFromAnyAmountOfNumbers()
+        {
+            var calculator = new StringCalculator();
+            var expected1 = 6;
+            var expected2 = 20;
+
+            var actual1 = calculator.Add("1,2,3");
+            var actual2 = calculator.Add("3,5,3,9");
```
Comment (mayankadlakha) [inline StringCalculator.Tests/StringCalculatorTests.cs:55]: Would be good to think of more descriptive variable names. @pnakkala is there a better way to do multiple assertions under one test name?

**Review by mayankadlakha — COMMENTED** [review]
Overall, really good unit tests. You have got descriptive method names and structured test cases (expected, actual, assertion). Also, this implementation seems quite tricky with a lot of string manipulation, so great work on that too 😀

File: StringCalculator.Tests/StringCalculatorTests.cs (hunk 8 above)
Comment (pnakkala) [reply StringCalculator.Tests/StringCalculatorTests.cs:55]: I have heard somewhere people saying a unit test method should only have one assertion. In practice, i have seen multiple assertions under one test method

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
+using System;
+using Xunit;
+
+namespace StringCalculator.Tests
+{
+    public class StringCalculatorShould
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:6]: `Should` at the end of the class name is not normal. Usually, it will be class_nameTest
e.g `StringCalculatorTest`
When you don't have `Should` in the class name, Then test methods will read like `Returns X when Y` format

File: StringCalculator.Tests/StringCalculatorTests.cs (hunk 8 above)
Comment (pnakkala) [reply StringCalculator.Tests/StringCalculatorTests.cs:55]: That said, current name seems alright to me

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+            var actual1 = calculator.Add("1,2,3");
+            var actual2 = calculator.Add("3,5,3,9");
+
+            Assert.Equal(expected1, actual1);
+            Assert.Equal(expected2, actual2);
+        }
+
+        [Fact]
+        public void AcceptNewLineBreaksAndCommas()
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:62]: Could even split into two tests. `NewLineBreak` `commas`

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        public void SupportDifferentDelimiters()
+        {
+            var calculator = new StringCalculator();
+            var expectedDelimiter = ";";
+            var expected = 3;
+
+            var actualDelimiter = calculator.CreateDelimiterList("//;\n1;2")[2];
+            var actual = calculator.Add("//;\n1;2");
+
+            Assert.Equal(expectedDelimiter, actualDelimiter);
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:85]: This assertion is not a valid test because we don't need to test this. Only `add` functionality

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.Equal(expected, actual);
+        }
+
+        [Fact]
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:89]: There may be a way to just annotate this method as `@Throws(ArgumentException)`
not sure if `xunit` has something like this. Then the code inside test method becomes as simple as
just one line `calculator.Add("-1,2,-3");`
I can see one downside of doing that, we won't be able to test the message of the exception. But it's good to know different ways of testing if an exception is raised

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+            var calculator = new StringCalculator();
+            var expected = 10;
+
+            var actual = calculator.Add("//[***][#][%]\n1***2#3%4");
+
+            Assert.Equal(expected, actual);
+        }
+
+        [Fact]
+        public void HandleDelimitersThatIncludeNumbers()
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:152]: We could write one negative scenario where digit appears at the edge of delimiters like "//[1**][%]\n1*1*2%3"
I'd imagine it would throw an exception

File: StringCalculator.Tests/StringCalculatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            var calculator = new StringCalculator();
+            var expected = 6;
+
+            var actual = calculator.Add("//[*1*][%]\n1*1*2%3");
+
+            Assert.Equal(expected, actual);
```
Comment (pnakkala) [inline StringCalculator.Tests/StringCalculatorTests.cs:159]: We could try to make use of more methods from xunit. Currently, there is only `Assert.Equal` method used. I think in this example there is not much scope to use many methods. So that's alright. Could explore if there is any method that asserts exceptions.

File: StringCalculator/StringCalculator.cs (hunk 4 above)
Comment (pnakkala) [reply StringCalculator/StringCalculator.cs:13]: yeah. static may be a good candidate here. We also need to be mindful of testing static methods. Many testing libraries do not support that. Does xunit support @mwGHburger

Comment (pnakkala) [reply StringCalculator/StringCalculator.cs:13]: yeah, static may be a good candidate here. If we do that, this class can become `final` as it will be utility class and we don't want anyone to extend it

File: StringCalculator/StringCalculator.cs [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            var total = 0;
+            if (String.IsNullOrEmpty(stringInput)) return total;
+
+            if (stringInput.Contains("//"))
```
Comment (pnakkala) [inline StringCalculator/StringCalculator.cs:18]: "//" can be extracted as a constant. That way if this constant is used in many places. you only need to change it once, when you need to

File: StringCalculator/StringCalculator.cs [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+            if (negativeStringNumbers.Count > 0)
+            {
+                throw new ArgumentException($"Throws exception with Negatives not allowed: {String.Join(", ", negativeStringNumbers)}");
+            }
+
+            return total;
+        }
+
+        public List<string> CreateDelimiterList(string text = "")
```
Comment (pnakkala) [inline StringCalculator/StringCalculator.cs:54]: could be `private`

**Comment by pnakkala** [comment]
Overall, You did a great job on TDD. keep going  @mwGHburger

**Comment by pnakkala** [comment]
By the way, i see that there are 131 files. only a few of them are relevant to review. What are those remaining files?

File: StringCalculator.Tests/StringCalculatorTests.cs (hunk 12 above)
Comment (mwGHburger) [reply StringCalculator.Tests/StringCalculatorTests.cs:89]: Will need to investigate annotations, but I see your point
